pkg/runner/test_runner_route.go

```diff
@@ -171,125 +180,148 @@ func (r *stepRouter) emitOutEvent(t *target.Target, err error) error {
+				log.Debugf("step error closed")
+				r.routingChannels.stepErr = nil
+				break
+			}
+
+			if _, targetPresent := egressTarget[targetError.Target]; targetPresent {
+				err = fmt.Errorf("step %s returned target %+v multiple times", r.bundle.TestStepLabel, targetError.Target)
+			} else {
+				if err := r.emitOutEvent(targetError.Target, targetError.Err); err != nil {
+					log.Warningf("could not emit err event for target: %v", *targetError.Target)
+				}
+				egressTarget[targetError.Target] = time.Now()
+				if err := targetWriter.writeTargetError(terminate, r.routingChannels.targetErr, targetError, r.timeouts.MessageTimeout); err != nil {
+					log.Panicf("could not forward target (%+v) to the test runner: %v", targetError.Target, err)
+				}
+			}
+		}
+		if err != nil {
+			break
+		}
+		if r.routingChannels.stepErr == nil && r.routingChannels.stepOut == nil {
+			log.Debugf("output and error channel from step are closed, routeOut should terminate")
+			close(r.routingChannels.routeOut)
+			break
+		}
+	}
 
-}
+	if err != nil {
+		log.Debugf("routeOut failed: %v", err)
+		return 0, err
+	}
+	return len(egressTarget), nil
 
+}
 
 // route implements the routing logic from the previous routing block to the test step
 // and from the test step to the next routing block
 func (r *stepRouter) route(terminate <-chan struct{}, resultCh chan<- routeResult) {
 
-        var (
-                inTargets, outTargets int
-                inErr, outErr, err    error
-                routeWg               sync.WaitGroup
-        )
-
-        routeWg.Add(1)
-        go func() {
-                defer routeWg.Done()
-                inTargets, inErr = r.routeIn(terminate)
-        }()
-
-        routeWg.Add(1)
-        go func() {
-                defer routeWg.Done()
-                outTargets, outErr = r.routeOut(terminate)
-        }()
-
-        routeWg.Wait()
-        if inErr != nil {
-                err = inErr
-        }
-        if outErr != nil {
-                err = outErr
-        }
-
-        if err == nil && inTargets != outTargets {
-                err = fmt.Errorf("step %s completed but did not return all injected Targets (%d!=%d)", r.bundle.TestStepLabel, inTargets, outTargets)
-        }
-
-        // Send the result to the test runner, which is expected to be listening
-        // within `MessageTimeout`. If that's not the case, we hit an unrecovrable
-        // condition.
-        select {
-        case resultCh <- routeResult{bundle: r.bundle, err: err}:
-        case <-time.After(r.timeouts.MessageTimeout):
-                log.Panicf("could not send routing block result")
-        }
+	var (
+		inTargets, outTargets               int
+		routingErr                          error
+		routeInCompleted, routeOutCompleted bool
+	)
+	terminateInternal := make(chan struct{})
+	errRouteInCh := make(chan error)
+	errRouteOutCh := make(chan error)
+
+	go func() {
+		var err error
+		inTargets, err = r.routeIn(terminateInternal)
```

Comment:
Why is there a terminate internal? You could just pass terminate through to every one of these.

Reply:
My intention is for `terminate` to be owned by the caller, which shouldn't have to bother to check if has been closed already. So only the owner can close it.

Reply:
Our rationale behind moving away from context is more or less aligned with https://dave.cheney.net/2017/08/20/context-isnt-for-cancellation. We were planning to move to https://github.com/insomniacslk/interruption, but never happened.